torchdynamo_poc/TorchDynamo Demo.ipynb

```diff
@@ -309,29 +378,79 @@
       "  func.func @forward(%arg0: !torch.vtensor<[2,3],f32>, %arg1: !torch.vtensor<[2,3],f32>) -> (!torch.vtensor<[2,3],f32>, !torch.vtensor<[],i1>) {\n",
       "    %int0 = torch.constant.int 0\n",
-      "    %int1 = torch.constant.int 1\n",
-      "    %none = torch.constant.none\n",
-      "    %0 = torch.aten.add.Scalar %arg0, %int1, %int1 : !torch.vtensor<[2,3],f32>, !torch.int, !torch.int -> !torch.vtensor<[2,3],f32>\n",
-      "    %1 = torch.aten.div.Tensor %arg0, %0 : !torch.vtensor<[2,3],f32>, !torch.vtensor<[2,3],f32> -> !torch.vtensor<[2,3],f32>\n",
-      "    %2 = torch.aten.sum %arg1, %none : !torch.vtensor<[2,3],f32>, !torch.none -> !torch.vtensor<[],f32>\n",
-      "    %3 = torch.aten.lt.Scalar %2, %int0 : !torch.vtensor<[],f32>, !torch.int -> !torch.vtensor<[],i1>\n",
-      "    return %1, %3 : !torch.vtensor<[2,3],f32>, !torch.vtensor<[],i1>\n",
+      "    %int2 = torch.constant.int 2\n",
+      "    %0 = torch.aten.div.Scalar %arg0, %int2 : !torch.vtensor<[2,3],f32>, !torch.int -> !torch.vtensor<[2,3],f32>\n",
+      "    %1 = torch.aten.max %arg1 : !torch.vtensor<[2,3],f32> -> !torch.vtensor<[],f32>\n",
+      "    %2 = torch.aten.lt.Scalar %1, %int0 : !torch.vtensor<[],f32>, !torch.int -> !torch.vtensor<[],i1>\n",
+      "    return %0, %2 : !torch.vtensor<[2,3],f32>, !torch.vtensor<[],i1>\n",
       "  }\n",
       "}\n",
       "\n",
+      "Compiling graph...\n",
+      "torch.fx graph:\n",
+      "graph():\n",
+      "    %arg0_1 : [#users=1] = placeholder[target=arg0_1]\n",
+      "    %arg1_1 : [#users=1] = placeholder[target=arg1_1]\n",
+      "    %mul : [#users=1] = call_function[target=torch.ops.aten.mul](args = (%arg1_1, %arg0_1), kwargs = {})\n",
+      "    return mul\n",
+      "\n",
+      "\n",
+      "torch-mlir backend contract graph:\n",
+      "module attributes {torch.debug_module_name = \"_lambda\"} {\n",
+      "  func.func @forward(%arg0: !torch.vtensor<[2,3],f32>, %arg1: !torch.vtensor<[2,3],f32>) -> !torch.vtensor<[2,3],f32> {\n",
+      "    %0 = torch.aten.mul.Tensor %arg1, %arg0 : !torch.vtensor<[2,3],f32>, !torch.vtensor<[2,3],f32> -> !torch.vtensor<[2,3],f32>\n",
+      "    return %0 : !torch.vtensor<[2,3],f32>\n",
+      "  }\n",
+      "}\n",
+      "\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "tensor([[0.5000, 0.5000, 0.5000],\n",
+       "        [0.5000, 0.5000, 0.5000]])"
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "foo(torch.ones((2, 3)), torch.ones((2, 3)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "349c7526",
+   "metadata": {},
+   "source": [
+    "Notice that because the body of the `if` statement is not used, it is not compiled. If we change the inputs so that the body of the `if` statement is needed, then TorchDynamo will reuse the compiled graphs for the other parts of the function and only compile the body of the `if` statement."
```

Comment:
Am I understanding correctly that even if you pass in another set of floats with the same dtype and shape it will only re-trace if the actual values would cause the if branch to be taken?

Reply:
That's correct. Only the path taken by the input tensors is the path compiled. If you change the input such that a new path is needed, the compiler will be run on that new path.